plugin/code_actions.py

```diff
@@ -1,56 +1,124 @@
+            else:
+                code_action.hide_bulb()
+
+        if callback is not None:
+            callback(response)
+
+    def show_bulb(self) -> None:
+        region = self.view.sel()[0]
+        flags = sublime.DRAW_NO_FILL | sublime.DRAW_NO_OUTLINE
+        self.view.add_regions('lsp_bulb', [region], 'markup.changed', 'Packages/LSP/icons/lightbulb.png', flags)
+
+    def hide_bulb(self) -> None:
+        self.view.erase_regions('lsp_bulb')
 
-    def run(self, edit, event=None):
-        client = client_for_view(self.view)
-        if client:
-            pos = get_position(self.view, event)
-            row, col = self.view.rowcol(pos)
-            point_diagnostics = get_point_diagnostics(self.view, pos)
-            params = {
-                "textDocument": {
-                    "uri": filename_to_uri(self.view.file_name())
-                },
-                "context": {
-                    "diagnostics": list(diagnostic.to_lsp() for diagnostic in point_diagnostics)
-                }
+    def _get_code_action_params(self):
+        region = self.view.sel()[0]
+        pos = region.begin()
+        point_diagnostics = get_point_diagnostics(self.view, pos)
+        return {
+            "textDocument": {
+                "uri": filename_to_uri(self.view.file_name())
+            },
+            "range": region_to_range(self.view, region).to_lsp(),
+            "context": {
+                "diagnostics": list(diagnostic.to_lsp() for diagnostic in point_diagnostics)
             }
-            params["range"] = region_to_range(self.view, self.view.sel()[0]).to_lsp()
-            if event:  # if right-clicked, set cursor to menu position
-                sel = self.view.sel()
-                sel.clear()
-                sel.add(sublime.Region(pos))
+        }
+
+
+class LspCodeActionListener(sublime_plugin.ViewEventListener):
+    @classmethod
+    def is_applicable(cls, _settings):
+        if settings.show_bulb:
+            return True
+        return False
+
+    def on_selection_modified_async(self):
+        self.fire_request()
+
+    def on_modified_async(self):
+        self.handle_modified_async()
+
+    @debounce(0.8)
+    def handle_modified_async(self):
+        self.fire_request()
+
+    @debounce(0.5)
+    def fire_request(self):
+        self.view.run_command('lsp_update_code_actions')
 
-            client.send_request(Request.codeAction(params), self.handle_codeaction_response)
 
-    def handle_codeaction_response(self, response: 'List[Dict]') -> None:
+class LspUpdateCodeActionsCommand(sublime_plugin.TextCommand):
+    def run(self, edit):
+        self.code_action = CodeAction(self.view)
+        self.code_action.send_request()
+
+
+class LspCodeActionsCommand(LspTextCommand):
+    def is_enabled(self):
+        return self.has_client_with_capability('codeActionProvider')
+
+    def run(self, edit):
+        self.commands = []  # type: List[Dict]
+
+        self.code_action = CodeAction(self.view)
```

Comment:
I'm not entirely sure why the `CodeAction` class should exist. Why is it a bad idea to merge the CodeAction class and this LspCodeActionsCommand class?

Reply:
It is not a bad idea. 

But first let me tell you why it first exist. When I started this, my initial idea was to have a class CodeAction that is responsible for the state and other things related to code actions. That way I have one place to send the request, cache data, etc. 
CodeAction class would be the source of truth. And other classes that inherit sublime_plugin ViewEventListener, TextCommand ... could just ask CodeActions what they need and do their thing based on the data they received :D

Well that was the initial idea. But I have nothing against simplicity, so now that I have no caching, I see no need to have a CodeAction class :)

Reply:
@rwols I gave the feedback to keep `LspCodeActionsCommand` unchanged as it makes it easier to add feature-flag and reason about the bulb behaviour separately. We can always merge the two later.

Reply:
Sure!